src/github.com/matrix-org/dendrite/common/sql.go

```diff
@@ -66,3 +68,9 @@ func TxStmt(transaction *sql.Tx, statement *sql.Stmt) *sql.Stmt {
 	}
 	return statement
 }
+
+// IsUniqueConstraintViolationErr returns true if the error is a postgresql unique_violation error
+func IsUniqueConstraintViolationErr(err error) bool {
+	pqErr, ok := err.(*pq.Error)
+	return ok && pqErr.Code == "23505"
```

Comment:
This should be `ok && err.Code.Class() == "23"`, no?

Reply:
`23505` is the specific error code for `unique_violation`. I commented on this [here](https://github.com/matrix-org/dendrite/pull/372#discussion_r156749720) but it got hidden by the following commit

Reply:
Hmm, having dug around it does look like only that error code gets used for all the various ways of defining primary keys/unique constraints.